FEXCore/Source/Interface/Core/JIT/JITClass.h

```diff
@@ -315,12 +325,201 @@ class Arm64JITCore final : public CPUBackend, public Arm64Emitter {
   void EmitLinkedBranch(uint64_t GuestRIP, bool Call) {
     PendingJumpThunks.push_back({GetCursorAddress<uint64_t>(), GuestRIP, {}});
     auto& Thunk = PendingJumpThunks.back();
-    Bind(&Thunk.Label);
+    BindOrRestart(&Thunk.Label);
     if (Call) {
-      bl(&Thunk.Label);
+      bl_OrRestart(&Thunk.Label);
     } else {
-      b(&Thunk.Label);
+      b_OrRestart(&Thunk.Label);
+    }
+  }
+
+  // Restart helpers
+  template<typename T>
+  requires (std::is_same_v<T, ARMEmitter::ForwardLabel> || std::is_same_v<T, ARMEmitter::BackwardLabel> ||
+            std::is_same_v<T, ARMEmitter::BiDirectionalLabel> || std::is_same_v<T, ARMEmitter::ForwardLabel::Reference>)
+  void bl_OrRestart(T* Label) {
+    if (bl(Label)) [[likely]] {
+      return;
+    }
+
+    // We can support this but currently unnecessary.
+    LOGMAN_MSG_A_FMT("Tried to branch larger than 128MB away!");
+    // Restart(RestartOptions::RestartOptionControl::NeedsLongJumps);
+  }
+
+  template<typename T>
+  requires (std::is_same_v<T, ARMEmitter::ForwardLabel> || std::is_same_v<T, ARMEmitter::BackwardLabel> ||
+            std::is_same_v<T, ARMEmitter::BiDirectionalLabel> || std::is_same_v<T, ARMEmitter::ForwardLabel::Reference>)
+  void b_OrRestart(T* Label) {
+    if (b(Label)) [[likely]] {
+      return;
+    }
+
+    // We can support this but currently unnecessary.
+    LOGMAN_MSG_A_FMT("Tried to branch larger than 128MB away!");
+    // Restart(RestartOptions::RestartOptionControl::NeedsLongJumps);
+  }
+
+  template<typename T>
+  requires (std::is_same_v<T, ARMEmitter::ForwardLabel> || std::is_same_v<T, ARMEmitter::BackwardLabel> ||
+            std::is_same_v<T, ARMEmitter::BiDirectionalLabel> || std::is_same_v<T, ARMEmitter::ForwardLabel::Reference>)
+  void b_OrRestart(ARMEmitter::Condition Cond, T* Label) {
+    if (RestartControl.NeedsLongJumps) [[unlikely]] {
+      ARMEmitter::ForwardLabel Skip {};
+      // Invert check and skip.
+      (void)b(InvertCondition(Cond), &Skip);
+      if (!b(Label)) [[unlikely]] {
+        LOGMAN_MSG_A_FMT("Tried to branch larger than 128MB away!");
+      }
```

Comment:
Do I understand correctly that this is done here since unconditionally branches can do farther jumps than conditional ones?

Might be worth changing these comments to something like "Wrap a manual Cond check around an unconditional branch; this can encode larger offsets" then. The inverted Cond check itself seems self-explanatory to me.